### ISSUE TITLE
Consider supporting backwards compatiblity between `ophys` and `pophys`

### ISSUE THREAD

**Issue by bruno-f-cruz** [AllenNeuralDynamics/aind-data-schema-models#62]
A recent change made the previous tag `ophys` incompatible with the new `pophys` modality. In order to support previously serialized schemas, it would be nice to have some custom logic to coerce the previous definition to the current one.

This coercion mechanism should ideally make things compatible at the level of the python API (e.g. calling `Modality.OPHYS`) and at the level of deserialization. In both cases, the output should coerce to the new Modality.POPHYS object.

**Comment by dbirman** [comment]
Using the new codegen, if I just add Modality.OPHYS pointing to the _Pophys() model is that sufficient to support this?

**Comment by bruno-f-cruz** [comment]
Naively I would say no. Assuming I understood what you suggested,
from a technical point of view, the abbreviation map would somewhat break since only the abbreviation of the last entry would be added
```python
Modality.abbreviation_map = {m().abbreviation: m() for m in Modality.ALL}
```

At a more conceptual level, I also think that the biggest point is NOT making it compatible at the level of the python API but instead at the level of deserialization. In fact, I think we should slowly deprecate the use of OPHYS in the python API but support the deserialization of the OPHYS into POPHYS.

You can check the tests on the open PR to have a sense what cases I believe we should support. (https://github.com/AllenNeuralDynamics/aind-data-schema-models/pull/61/files#diff-a238cb580608af7c9b598c6b5bb3bacf973927b080ad7c8919e0ea104c1947cd)

**Comment by dbirman** [comment]
Got it thanks for the clarification -- in that case it might be easier to just run an upgrader on all the old metadata and replace the variable wherever we can find it. Unless there are places that OPHYS was used outside of the metadata but that need to be deserialized into one of the metadata classes, then we do need to implement as you've done here.

**Comment by dbirman** [comment]
Waiting on #94

**Comment by dbirman** [comment]
After poking at this a bit I'm not sure this is possible, the discriminated Union forces you to rely on the abbreviation field and after an hour of messing with this I can't find a way to get it to deserialize properly. The issue is that the `ONE_OF` field only has a limited list of options. You can get around this by adding an additional class with abbreviation="ophys" and you can get it to pick up that class during deserialization, but then if you run a model_validator on top of that to swap ophys->pophys it fails because it already knows the type of the class at that point and it expects abbreviation: Literal["ophys"].

One way around this would be to implement the validators in classes that include a call to Modality.ONE_OF (rig, session, quality_control, and data_description). I think this isn't worth it when we can go back and replace old uses of "ophys" -> "pophys". 

I'll open a ticket in aind-data-migration to migrate all previous uses of "ophys" -> "pophys" in the metadata.

If there's additional concerns with metadata that isn't stored in DocDB let me know
